### ISSUE TITLE
Generate CR Struct from CRD

### ISSUE THREAD

**Issue by samirketema** [volvo-cars/lingon#88]
Hi, this is a fantastic tool! 

I'm not the most experienced with Kubernetes and have been working on a problem to generate golang CR structs from a yaml file containing the CRD (from a third party library).

Using lingon, I was able to generate a variable containing the CRD. I am curious as to how I could instead generate a struct representing the custom resource itself, so I can create custom resources via the k8s API. I only have the yaml file representing the CRD.

**Comment by veggiemonk** [comment]
Hi @samirketema,

Thank you for the feedback, much appreciated.

Interesting question, if I understand correctly, you want to generate the definition of a custom resource (CRD) by just having a custom resource (CR), correct?

It is much more difficult to do because it is like generate a schema from a sample of data.
Another example would be to try to generate the OpenAPI spec from just a few requests.
It may or may not be correct because the sample won't probably contain all the possible values a schema can hold. 

However, it might be possible to find the actual schema. When importing the custom resource in Go, the generated Go files contain an import statement. There is a high chance that the types are defined in that package. That's where you would potentially find the CRDs, if you have access to the source, of course.

To learn more about CRDs, the kubebuilder book has an example of how to build such an API for custom resource https://book.kubebuilder.io/cronjob-tutorial/api-design.html 

I hope it helps

**Comment by samirketema** [comment]
Hi @veggiemonk, thank you for all the references the thoughtful response.

I may have misrepresented the problem - it's actually the other way around. I already have the CRD already defined in a yaml, and I am attempting to create instance of the custom resource, based on that definition. I would like to generate a struct representing an instance of the resource, rather than the resource definition itself.

I have already wrote a script doing this using go generate, but I feel that using like this tool is more robust, since there are many kubernetes types that this supports.

**Comment by veggiemonk** [comment]
Hi @samirketema, 

Sorry I misunderstood, it becomes very meta when it comes to generating code.
The best way to communicate about this is just to show code and commands. 

I do not know how to go from CRD to a valid go struct CR but I'd be curious to see what command you used with `go generate` to obtain such results. Maybe that's something we can add in the future. Can you show us, please?

**Comment by samirketema** [comment]
@veggiemonk no worries at all! absolutely agree, showing code is certainly more clear.

My script is extremely rough (needs a significant refactor) and is very heavily tied to a specific CRD `FlinkDeployment` from the [Apache Flink Kubernetes operator](https://github.com/apache/flink-kubernetes-operator/blob/main/helm/flink-kubernetes-operator/crds/flinkdeployments.flink.apache.org-v1.yml). Changes are required to generalize it to support other CRDs, fix some glaring issues, and improve readability/performance.  Not my proudest code, but I didn't have much time to improve it 😂 

to execute:
```bash
go generate ./...
```

generator script:

```go
//go:generate go run cr_generator.go
//go:generate go fmt ./flink_types.go
package main

// Generates a Custom Resource struct from the FlinkDeployment CRD yaml.

import (
	"fmt"
	"log"
	"os"
	"regexp"
	"sort"
	"text/template"
	"unicode"

	"gopkg.in/yaml.v2"
)

type Generator struct {
	Package      string
	GroupVersion string
	Structs      map[string]Struct
	Enums        map[string]Enum
}

type Enum struct {
	Name   string
	Values []string
}

type Struct struct {
	KubebuilderAnnotations []string
	Name                   string
	Fields                 []Field
	MetaTypes              []MetaType
}

type MetaType struct {
	Type string
	Tag  string
}

type Field struct {
	Name string
	Type string
	Tag  string
}

// TODO: use k8s runtime deserializer code to automtically detect built-in k8s types.
// TODO: write tests :sob:
func main() {
	// Read the input YAML file
	data, err := os.ReadFile("./input-crd.yaml")
	if err != nil {
		log.Fatal(err)
	}

	// Parse the YAML into a Generator struct
	var yamlContents map[string]interface{}
	err = yaml.Unmarshal(data, &yamlContents)
	if err != nil {
		log.Fatal(err)
	}

	// navigate the yaml to the portion containing the definition

	spec := yamlContents["spec"].(map[interface{}]interface{})

	// TODO: decrease memory footprint
	group := spec["group"].(string)

	names := spec["names"].(map[interface{}]interface{})
	kind := names["kind"].(string)

	versions := spec["versions"].([]interface{})
	version := versions[0].(map[interface{}]interface{})
	versionName := version["name"].(string)
	schema := version["schema"].(map[interface{}]interface{})
	openAPIV3Schema := schema["openAPIV3Schema"].(map[interface{}]interface{})
	schemaProperties := openAPIV3Schema["properties"].(map[interface{}]interface{})

	// Generate the Go code
	generateCode(kind, group, versionName, schemaProperties)
}

func generateCode(crdName, group, version string, yamlContents map[interface{}]interface{}) {
	// Create the output file
	file, err := os.Create("./flink_types.go")
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	generator := &Generator{
		Package:      "flink",
		GroupVersion: fmt.Sprintf(`GroupVersion = schema.GroupVersion{Group: "%s", Version: "%s"}`, group, version),
		Structs:      make(map[string]Struct),
		Enums:        make(map[string]Enum),
	}

	parseStructs("flinkDeployment", "", "metav1.ObjectMeta", yamlContents, generator, true)
	err = codeTemplate.Execute(file, generator)
	if err != nil {
		log.Panic("failed to execute code template:", err)
	}
}

// TODO: refactor. bring terminal cases to top for clarity
// TODO: make assumptions more clear.
// TODO: generalize this to make it compatible with other CRDs, support more k8s builtin types
func parseStructs(key, parentKey, structType string, m interface{}, g *Generator, properties bool) {
	// Titlecase the key, to keep struct fields and types public for now
	titleKey := capitalize(key)

	// check for collisions, try appending the parent key.
	if _, collision := g.Structs[titleKey]; collision {
		titleKey = capitalize(parentKey) + titleKey
	}

	// try again, if another collision, panic
	if _, collision := g.Structs[titleKey]; collision {
		log.Panicf("yaml collision on %s", titleKey)
	}

	s := Struct{
		Name:      titleKey,
		MetaTypes: []MetaType{},
	}

	// expect a map here
	mm, ok := m.(map[interface{}]interface{})
	if !ok {
		panic("unexpected structure")
	}

	// ignore type of the struct itself, we'll instead determine type from the struct fields
	delete(mm, "type")

	// iterate through map in order, to have a deterministic code generation
	// this is due to maps being unordered
	fieldKeys := make([]string, len(mm))
	i := 0
	for fk := range mm {
		fieldKeys[i] = fk.(string)
		i++
	}
	sort.Strings(fieldKeys)

	// iterate through fields
	for _, fieldName := range fieldKeys {
		fv := mm[fieldName]
		titleFieldName := capitalize(fieldName)

		// This should be a nested type
		fieldValue := fv.(map[interface{}]interface{})

		// before we look into all the fields, let's determine if this is an enum
		if e, ok := fieldValue["enum"].([]interface{}); ok {
			// add an enum
			enumValues := make([]string, len(e))
			for i, v := range e {
				enumValues[i] = v.(string)
			}
			addEnum(g, titleFieldName, enumValues)
			s.Fields = append(s.Fields, Field{
				Name: titleFieldName,
				Type: titleFieldName,
				Tag:  fmt.Sprintf("`json:\"%s,omitempty\"`", fieldName),
			})
			continue
		}

		// also, if it has an "additionalProperties" type, it's just a map
		if _, ok := fieldValue["additionalProperties"].(map[interface{}]interface{}); ok {
			// just add the map field and continue
			s.Fields = append(s.Fields, Field{
				Name: titleFieldName,
				Type: "map[string]string",
				Tag:  fmt.Sprintf("`json:\"%s,omitempty\"`", fieldName),
			})
			continue
		}

		if t, ok := fieldValue["type"].(string); ok && t != "" {
			// determine type
			determinedType := determineType(fieldName, t)

			outputType := determinedType
			switch determinedType {
			case "metav1.ObjectMeta":
				outputType = titleFieldName
			case "metav1.ListMeta":
				outputType = titleFieldName
			}

			f := Field{
				Name: titleFieldName,
				Type: outputType,
				Tag:  fmt.Sprintf("`json:\"%s,omitempty\"`", fieldName),
			}

			// we will need to recursively define a struct for nested objects and arrays
			switch determinedType {
			case "metav1.ObjectMeta":
				// it should either have a properties field
				if p, ok := fieldValue["properties"]; ok {
					parseStructs(fieldName, titleKey, determinedType, p, g, true)
				} else {
					// or an additional properties field
					_, ok := fieldValue["additionalProperties"]
					if !ok {
						log.Panicf("unexpected map, missing additional properties: %s", fieldName)
					}

					parseStructs(fieldName, titleKey, determinedType, fieldValue, g, false)
				}
			case "metav1.ListMeta":
				// it should have an items field, add a struct for it
				items, ok := fieldValue["items"].(map[interface{}]interface{})
				if !ok {
					log.Panicf("unexpected array type for field: %s", fieldName)
				}

				itemType := items["type"].(string)

				// if the item is an object, parse the underlying struct type
				switch itemType {
				case "object":
					parseStructs(fieldName, titleKey, determinedType, items["properties"], g, false)
				case "array":
					parseStructs(fieldName, titleKey, determinedType, items, g, false)
				default:
					// otherwise, change the field type to a slice
					f.Type = fmt.Sprintf("[]%s", itemType)
				}
			}

			// add the field to the struct
			s.Fields = append(s.Fields, f)
		}
	}

	addStruct(g, structType, s)
}

// This works for now, but if the FlinkDeployment includes other k8s types,
// we may miss out on them silently.
func determineType(key, t string) string {
	switch key {
	case "additionalProperties":
		return "map[string]string"
	case "podTemplate":
		return "corev1.PodTemplate"
	case "imagePullPolicy":
		return "corev1.PullPolicy"
	case "resources":
		return "corev1.ResourceRequirements"
	}

	switch t {
	case "string":
		return "string"
	case "object":
		return "metav1.ObjectMeta"
	case "array":
		return "metav1.ListMeta"
	case "integer":
		return "int64"
	case "number":
		return "float64"
	case "boolean":
		return "bool"
	default:
		return t
	}
}

func addStruct(g *Generator, structType string, s Struct) {
	s.KubebuilderAnnotations = []string{
		"// +kubebuilder:object:root=true",
		"// +groupName=flink.apache.org",
	}

	// Note the relevant metatypes
	switch structType {
	case "metav1.ObjectMeta":
		s.MetaTypes = []MetaType{
			{
				Type: "metav1.TypeMeta",
				Tag:  "`json:\",inline\"`",
			},
			{
				Type: "metav1.ObjectMeta",
				Tag:  "`json:\"metadata,omitempty\"`",
			},
		}
	case "metav1.ListMeta":
		s.MetaTypes = []MetaType{
			{
				Type: "metav1.TypeMeta",
				Tag:  "`json:\",inline\"`",
			},
			{
				Type: "metav1.ListMeta",
				Tag:  "`json:\"metadata,omitempty\"`",
			},
		}
	default:
		log.Panicf("unexpected struct type: type=%s", structType)
	}

	// add to generator
	g.Structs[s.Name] = s
}

func addEnum(g *Generator, enumName string, values []string) {
	g.Enums[enumName] = Enum{
		Name:   enumName,
		Values: values,
	}
}

func capitalize(str string) string {
	if str == "" {
		return str
	}
	runes := []rune(str)
	runes[0] = unicode.ToUpper(runes[0])
	return string(runes)
}

// sanitizes enum values so we can use them in constant names
func sanitize(s string) string {
	reg, err := regexp.Compile("[^a-zA-Z0-9]+")
	if err != nil {
		log.Fatal(err)
	}
	return reg.ReplaceAllString(s, "")
}

var funcMap = template.FuncMap{
	"capitalize": capitalize,
	"sanitize":   sanitize,
}

var codeTemplate = template.Must(template.New("").Funcs(funcMap).Parse(`// Code generated by go generate; DO NOT EDIT.

package {{ .Package }}

import (
	corev1 "k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/runtime/schema"
	"sigs.k8s.io/controller-runtime/pkg/scheme"
)

var (
	// GroupVersion is group version used to register these objects
	{{ .GroupVersion }}

	// SchemeBuilder is used to add go types to the GroupVersionKind scheme
	SchemeBuilder = &scheme.Builder{GroupVersion: GroupVersion}

	// AddToScheme adds the types in this group-version to the given scheme.
	AddToScheme = SchemeBuilder.AddToScheme
)

func init() {
	SchemeBuilder.Register(&FlinkDeployment{})
}

{{ range $e := .Enums}}

type {{ $e.Name }} string

const (
{{- range $v := $e.Values }}
{{- $s := sanitize $v }}
{{ $e.Name }}{{ capitalize $s }} = "{{ $v }}"
{{- end }}
)
{{ end }}

{{ range .Structs}}
	{{- range $a := .KubebuilderAnnotations}}
	{{ $a }}
	{{- end}}

	type {{ .Name }} struct {
		{{ range .MetaTypes }} {{ .Type }} {{ .Tag }}
		{{ end }}
		{{ range .Fields }}	{{ .Name }} {{ .Type }} {{ .Tag }}
		{{ end }}
	}
{{ end }}
`))

```
and here is the resulting `FlinkDeployment` custom resource generated from that script:

```go
// Code generated by go generate; DO NOT EDIT.

package flink

import (
	corev1 "k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/runtime/schema"
	"sigs.k8s.io/controller-runtime/pkg/scheme"
)

var (
	// GroupVersion is group version used to register these objects
	GroupVersion = schema.GroupVersion{Group: "flink.apache.org", Version: "v1beta1"}

	// SchemeBuilder is used to add go types to the GroupVersionKind scheme
	SchemeBuilder = &scheme.Builder{GroupVersion: GroupVersion}

	// AddToScheme adds the types in this group-version to the given scheme.
	AddToScheme = SchemeBuilder.AddToScheme
)

func init() {
	SchemeBuilder.Register(&FlinkDeployment{})
}

type FlinkVersion string

const (
	FlinkVersionV113 = "v1_13"
	FlinkVersionV114 = "v1_14"
	FlinkVersionV115 = "v1_15"
	FlinkVersionV116 = "v1_16"
	FlinkVersionV117 = "v1_17"
)

type FormatType string

const (
	FormatTypeCANONICAL = "CANONICAL"
	FormatTypeNATIVE    = "NATIVE"
	FormatTypeUNKNOWN   = "UNKNOWN"
)

type JobManagerDeploymentStatus string

const (
	JobManagerDeploymentStatusREADY            = "READY"
	JobManagerDeploymentStatusDEPLOYEDNOTREADY = "DEPLOYED_NOT_READY"
	JobManagerDeploymentStatusDEPLOYING        = "DEPLOYING"
	JobManagerDeploymentStatusMISSING          = "MISSING"
	JobManagerDeploymentStatusERROR            = "ERROR"
)

type LifecycleState string

const (
	LifecycleStateCREATED     = "CREATED"
	LifecycleStateSUSPENDED   = "SUSPENDED"
	LifecycleStateUPGRADING   = "UPGRADING"
	LifecycleStateDEPLOYED    = "DEPLOYED"
	LifecycleStateSTABLE      = "STABLE"
	LifecycleStateROLLINGBACK = "ROLLING_BACK"
	LifecycleStateROLLEDBACK  = "ROLLED_BACK"
	LifecycleStateFAILED      = "FAILED"
)

type Mode string

const (
	ModeNative     = "native"
	ModeStandalone = "standalone"
)

type State string

const (
	StateDEPLOYED    = "DEPLOYED"
	StateUPGRADING   = "UPGRADING"
	StateROLLINGBACK = "ROLLING_BACK"
	StateROLLEDBACK  = "ROLLED_BACK"
)

type TriggerType string

const (
	TriggerTypeMANUAL   = "MANUAL"
	TriggerTypePERIODIC = "PERIODIC"
	TriggerTypeUPGRADE  = "UPGRADE"
	TriggerTypeUNKNOWN  = "UNKNOWN"
)

type UpgradeMode string

const (
	UpgradeModeSavepoint = "savepoint"
	UpgradeModeLaststate = "last-state"
	UpgradeModeStateless = "stateless"
)

// +kubebuilder:object:root=true
// +groupName=flink.apache.org

type FlinkDeployment struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	Spec   Spec   `json:"spec,omitempty"`
	Status Status `json:"status,omitempty"`
}

// +kubebuilder:object:root=true
// +groupName=flink.apache.org

type Ingress struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	Annotations map[string]string `json:"annotations,omitempty"`
	ClassName   string            `json:"className,omitempty"`
	Template    string            `json:"template,omitempty"`
}

// +kubebuilder:object:root=true
// +groupName=flink.apache.org

type Job struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	AllowNonRestoredState bool        `json:"allowNonRestoredState,omitempty"`
	Args                  []string    `json:"args,omitempty"`
	EntryClass            string      `json:"entryClass,omitempty"`
	InitialSavepointPath  string      `json:"initialSavepointPath,omitempty"`
	JarURI                string      `json:"jarURI,omitempty"`
	Parallelism           int64       `json:"parallelism,omitempty"`
	SavepointTriggerNonce int64       `json:"savepointTriggerNonce,omitempty"`
	State                 State       `json:"state,omitempty"`
	UpgradeMode           UpgradeMode `json:"upgradeMode,omitempty"`
}

// +kubebuilder:object:root=true
// +groupName=flink.apache.org

type JobManager struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	PodTemplate corev1.PodTemplate `json:"podTemplate,omitempty"`
	Replicas    int64              `json:"replicas,omitempty"`
	Resource    Resource           `json:"resource,omitempty"`
}

// +kubebuilder:object:root=true
// +groupName=flink.apache.org

type JobStatus struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	JobId         string        `json:"jobId,omitempty"`
	JobName       string        `json:"jobName,omitempty"`
	SavepointInfo SavepointInfo `json:"savepointInfo,omitempty"`
	StartTime     string        `json:"startTime,omitempty"`
	State         string        `json:"state,omitempty"`
	UpdateTime    string        `json:"updateTime,omitempty"`
}

// +kubebuilder:object:root=true
// +groupName=flink.apache.org

type LastSavepoint struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	FormatType   FormatType  `json:"formatType,omitempty"`
	Location     string      `json:"location,omitempty"`
	TimeStamp    int64       `json:"timeStamp,omitempty"`
	TriggerNonce int64       `json:"triggerNonce,omitempty"`
	TriggerType  TriggerType `json:"triggerType,omitempty"`
}

// +kubebuilder:object:root=true
// +groupName=flink.apache.org

type ReconciliationStatus struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	LastReconciledSpec      string `json:"lastReconciledSpec,omitempty"`
	LastStableSpec          string `json:"lastStableSpec,omitempty"`
	ReconciliationTimestamp int64  `json:"reconciliationTimestamp,omitempty"`
	State                   State  `json:"state,omitempty"`
}

// +kubebuilder:object:root=true
// +groupName=flink.apache.org

type Resource struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	Cpu              float64 `json:"cpu,omitempty"`
	EphemeralStorage string  `json:"ephemeralStorage,omitempty"`
	Memory           string  `json:"memory,omitempty"`
}

// +kubebuilder:object:root=true
// +groupName=flink.apache.org

type SavepointHistory struct {
	metav1.TypeMeta `json:",inline"`
	metav1.ListMeta `json:"metadata,omitempty"`

	FormatType   FormatType  `json:"formatType,omitempty"`
	Location     string      `json:"location,omitempty"`
	TimeStamp    int64       `json:"timeStamp,omitempty"`
	TriggerNonce int64       `json:"triggerNonce,omitempty"`
	TriggerType  TriggerType `json:"triggerType,omitempty"`
}

// +kubebuilder:object:root=true
// +groupName=flink.apache.org

type SavepointInfo struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	FormatType                     FormatType       `json:"formatType,omitempty"`
	LastPeriodicSavepointTimestamp int64            `json:"lastPeriodicSavepointTimestamp,omitempty"`
	LastSavepoint                  LastSavepoint    `json:"lastSavepoint,omitempty"`
	SavepointHistory               SavepointHistory `json:"savepointHistory,omitempty"`
	TriggerId                      string           `json:"triggerId,omitempty"`
	TriggerTimestamp               int64            `json:"triggerTimestamp,omitempty"`
	TriggerType                    TriggerType      `json:"triggerType,omitempty"`
}

// +kubebuilder:object:root=true
// +groupName=flink.apache.org

type Spec struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	FlinkConfiguration map[string]string  `json:"flinkConfiguration,omitempty"`
	FlinkVersion       FlinkVersion       `json:"flinkVersion,omitempty"`
	Image              string             `json:"image,omitempty"`
	ImagePullPolicy    corev1.PullPolicy  `json:"imagePullPolicy,omitempty"`
	Ingress            Ingress            `json:"ingress,omitempty"`
	Job                Job                `json:"job,omitempty"`
	JobManager         JobManager         `json:"jobManager,omitempty"`
	LogConfiguration   map[string]string  `json:"logConfiguration,omitempty"`
	Mode               Mode               `json:"mode,omitempty"`
	PodTemplate        corev1.PodTemplate `json:"podTemplate,omitempty"`
	RestartNonce       int64              `json:"restartNonce,omitempty"`
	ServiceAccount     string             `json:"serviceAccount,omitempty"`
	TaskManager        TaskManager        `json:"taskManager,omitempty"`
}

// +kubebuilder:object:root=true
// +groupName=flink.apache.org

type Status struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	ClusterInfo                map[string]string          `json:"clusterInfo,omitempty"`
	Error                      string                     `json:"error,omitempty"`
	JobManagerDeploymentStatus JobManagerDeploymentStatus `json:"jobManagerDeploymentStatus,omitempty"`
	JobStatus                  JobStatus                  `json:"jobStatus,omitempty"`
	LifecycleState             LifecycleState             `json:"lifecycleState,omitempty"`
	ReconciliationStatus       ReconciliationStatus       `json:"reconciliationStatus,omitempty"`
	TaskManager                TaskManager                `json:"taskManager,omitempty"`
}

// +kubebuilder:object:root=true
// +groupName=flink.apache.org

type StatusTaskManager struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	LabelSelector string `json:"labelSelector,omitempty"`
	Replicas      int64  `json:"replicas,omitempty"`
}

// +kubebuilder:object:root=true
// +groupName=flink.apache.org

type TaskManager struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	PodTemplate corev1.PodTemplate `json:"podTemplate,omitempty"`
	Replicas    int64              `json:"replicas,omitempty"`
	Resource    Resource           `json:"resource,omitempty"`
}

// +kubebuilder:object:root=true
// +groupName=flink.apache.org

type TaskManagerResource struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	Cpu              float64 `json:"cpu,omitempty"`
	EphemeralStorage string  `json:"ephemeralStorage,omitempty"`
	Memory           string  `json:"memory,omitempty"`
}

```

**Comment by samirketema** [comment]
@veggiemonk to add more context for the use case that prompted me to write that script, now that I have that `FlinkDeployment` CR struct type created, I can dynamically perform CRUD operations with instances of that resource, directly in go code

**Comment by veggiemonk** [comment]
Alright @samirketema ! I see, thanks for clarifying. 

There is a lot to unpack here. 

First, I'm still not 100% sure about the *why* you are doing this as there are already a few Flink operators out there, https://github.com/search?utf8=%E2%9C%93&q=flink+operator+language:Go&type=repositories&l=Go
Copying one of those into your code base is more common than I thought.
Istio for example, copy the kubernetes types so they don't need to import the dependencies.

Copying and potentially adapting those structs to the use case might save a bit of time. That is, if your solution already work, don't fix it if it is not broken. 

Now, when it comes to the definitions (or schema) of a custom resource: 
Taking the CRD from yaml to GO and then creating a struct (called CR or custom resource) based on that CRD seems quite tedious but it works! Although, generalizing that to every CRDs seems like a daunting task and it might take a huge effort to fix all edge cases and bugs. So this would not be my first approach. Basically, building a "reverse kubebuilder" sounds like a lot of more work. 

However, CRDs are mostly OpenAPIv3 schemas and that is very well supported by a big ecosystem: https://github.com/deepmap/oapi-codegen

If I had to generate CR from a CRD, I would try that route as it will lead to less maintenance in the long run and more flexibility as you would be dealing with a standard schema format. 

The process would be CRD yaml -> OpenAPIV3 -> Go struct. 

I think you are on the right track as the code looks similar to your own: https://github.com/deepmap/oapi-codegen/blob/master/pkg/codegen/schema.go#L221

Let me know if it works

**Comment by samirketema** [comment]
@veggiemonk Thank you so much for the thorough answer, even though much of this is tangential to your project!

As for the why: I took a look at some of these operators and my initial assumption was that they extended or adjusted the behavior/architecture of the operator. I haven't had enough time to completely validate that assumption, but I will absolutely do so, since it can simplify the solution. Aside from the deprecated library, the Lyft and Spotify operators do seem like great alternatives.

I also weighed the pros/cons of copying & adapting the structs by hand and decided against it due to the need to update the struct over time due to changes upon upgrades.

I'll take a closer look into the OpenAPIv3 code generation. That sounds like a more robust approach. Thank you for the suggestion and your attention to detail!
